In [1]:
import numpy as np
import vectfit as m
import time

In [2]:
%load_ext memory_profiler

In [3]:
%memit

peak memory: 48.47 MiB, increment: -1.95 MiB


In [4]:
for N in [10, 100, 200, 500, 1000, 2000, 4000]:
    for Ns in [1000, 5000, 10000, 50000, 100000]:
        if Ns < 2*N:
            continue
        s = np.linspace(1.e-2, 5.e3, Ns)
        poles = np.linspace(1.1e-2, 4.8e+3, N//2);
        poles = poles + poles*0.01j
        poles = np.sort(np.append(poles, np.conj(poles)))
        residues = np.linspace(1e+2, 1e+6, N//2);
        residues = residues + residues*0.5j
        residues = np.sort(np.append(residues, np.conj(residues)))
        residues = residues.reshape((1, N))
        f = np.zeros((1, Ns))
        for p, r in zip(poles, residues[0]):
            f[0] += (r/(s - p)).real
        weight = 1.0/f
        poles_init = np.linspace(1.2e-2, 4.7e+3, N//2);
        poles_init = poles_init + poles_init*0.01j
        poles_init = np.sort(np.append(poles_init, np.conj(poles_init)))
        print("N={}, Ns={}:".format(N, Ns))
        tstart = time.time()
        %memit
        %memit poles_fit, residues_fit, cf, f_fit, rms = m.vectfit(f, s, poles_init, weight)
        %memit
        print("time: {:.2f} s\n".format(time.time() - tstart))

N=10, Ns=1000:
peak memory: 49.00 MiB, increment: 0.02 MiB
peak memory: 65.45 MiB, increment: 16.45 MiB
peak memory: 63.34 MiB, increment: -2.11 MiB
time: 0.58 s

N=10, Ns=5000:
peak memory: 63.51 MiB, increment: 0.17 MiB
peak memory: 67.72 MiB, increment: 4.21 MiB
peak memory: 62.07 MiB, increment: -2.11 MiB
time: 0.62 s

N=10, Ns=10000:
peak memory: 62.37 MiB, increment: 0.01 MiB
peak memory: 75.51 MiB, increment: 13.14 MiB
peak memory: 62.34 MiB, increment: -2.11 MiB
time: 0.69 s

N=10, Ns=50000:
peak memory: 64.59 MiB, increment: 0.00 MiB
peak memory: 125.99 MiB, increment: 61.41 MiB
peak memory: 64.20 MiB, increment: -2.11 MiB
time: 0.97 s

N=10, Ns=100000:
peak memory: 68.02 MiB, increment: 0.00 MiB
peak memory: 202.59 MiB, increment: 134.57 MiB
peak memory: 98.54 MiB, increment: -2.11 MiB
time: 1.44 s

N=100, Ns=1000:
peak memory: 98.55 MiB, increment: 0.00 MiB
peak memory: 128.94 MiB, increment: 30.39 MiB
peak memory: 90.75 MiB, increment: -1.63 MiB
time: 0.77 s

N=100, Ns=5000

| Memory(MiB) |      | # poles |         |         |         |         |         |
|-------------|------|---------|---------|---------|---------|---------|---------|
|             |      | 100     | 200     | 500     | 1000    | 2000    | 4000    |
| # points    | 1e3  | 3.0E+01 | 3.2E+01 | 1.4E+01 |         |         |         |
|             | 5e3  | 8.3E+01 | 1.1E+02 | 3.5E+02 | 6.9E+02 | 1.6E+03 |         |
|             | 1e4  | 1.2E+02 | 3.1E+02 | 6.9E+02 | 1.4E+03 | 3.0E+03 | 6.5E+03 |
|             | 5e4  | 6.9E+02 | 1.4E+03 | 3.4E+03 | 6.9E+03 | 1.4E+04 | 2.9E+04 |
|             | 1e5  | 1.4E+03 | 2.8E+03 | 6.9E+03 | 1.4E+04 | 2.8E+04 | 5.6E+04 |


| Time(s)  |      | # poles |       |       |        |        |         |
|----------|------|---------|-------|-------|--------|--------|---------|
|          |      | 100     | 200   | 500   | 1000   | 2000   | 4000    |
| # points | 1e3  | 0.77    | 1.04  | 1.78  |        |        |         |
|          | 5e3  | 0.97    | 1.59  | 4.59  | 14.90  | 42.33  |         |
|          | 1e4  | 1.52    | 2.96  | 8.78  | 24.52  | 76.81  | 279.03  |
|          | 5e4  | 5.29    | 12.35 | 36.78 | 105.65 | 320.35 | 1173.40 |
|          | 1e5  | 12.11   | 28.79 | 94.15 | 235.76 | 789.75 | 2997.20 |

+ Time ∝ N K
+ Memory ∝ (N K)^2


### U238:

**N~6000, Ns~6e5 (0.05% accuracy) -> memory ~ 500 GiB, Time > 24 h**

If splitting to 100 domains:

**N~100, Ns~1e4 -> memory ~ 100 MiB, Time ~ 10 s**


### U235:

**N~5000, Ns~3e5 (0.05% accuracy) -> memory ~ 200 GiB, Time > 12 h**

If splitting to 100 domains:

**N~100, Ns~5e3 -> memory ~ 100 MiB, Time ~ 5 s**